src/Tree.elm

```diff
@@ -0,0 +1,51 @@
+module Tree exposing (Tree(..), buildTree, name)
+
+import Data.Audience exposing (Audience)
+import Data.AudienceFolder exposing (AudienceFolder)
+import Dict
+import Dict.Extra exposing (groupBy)
+
+
+type Tree
+    = Tree AudienceFolder (List Tree) (List Audience)
+
+
+name (Tree f _ _) =
+    f.name
+
+
+makeBranch : AudienceFolder -> Dict.Dict Int (List AudienceFolder) -> Dict.Dict Int (List Audience) -> Tree
+makeBranch folder subfoldersDict leafDict =
+    let
+        getCurrentFolderData dict =
+            dict |> Dict.get folder.id |> Maybe.withDefault []
+
+        subfolders =
+            getCurrentFolderData subfoldersDict
+
+        leafs =
+            getCurrentFolderData leafDict
+    in
+        Tree folder (List.map (\f -> makeBranch f subfoldersDict leafDict) subfolders) leafs
```

Comment:
this recursion can blow stack with some data. On the other hand it’s quite common issue in situations like this as writing tail recursive variant is incredibely complex (and often has other performance trade offs).